docs/cookbook/rag.md

```diff
@@ -0,0 +1,257 @@
+### Pinecone Setup
+
+For a large dataset, we can speed up querying with a vector database service like [Pinecone](https://www.pinecone.io/). We set up a Pinecone index and upsert our embedded vectors to it in batches to keep requests small. Note that this time around, we specify cosine similarity in our index to account for the different lengths of each text in our database.
+
+```python
+pc = Pinecone(api_key="YOUR_API_KEY")
+
+pc.create_index(
+    name="news-articles",
+    dimension=1536,
+    metric="cosine",
+    spec=ServerlessSpec(
+        cloud="aws",
+        region="us-west-2",
+    ),
+)
+
+index = pc.Index(index_name)
+
+batch_size = 300  # Pinecone can't handle too large of an upsert request
+for start in range(0, len(df), batch_size):
+    end = min(start + batch_size, len(df))
+    batch = df[start:end]
+    vectors = []
+    for i, row in batch.iterrows():
+        vectors.append({"id": str(i), "values": row[EMBEDDINGS_COLUMN]})
+    index.upsert(vectors=vectors, namespace="articles")
+```
+
+### Final Product
+
+We're almost there! Just like the soccer example, we set up a `Prompt` to handle the chat structure. This time, we make our prompt a bit more complex, taking advantage of Mirascope's Pydantic prompts to nicely format the retrieved articles.
+
+```python
+class NewsPrompt(Prompt):
+    """
+    Here are {num_articles} relevant sections of news articles relevant to this topic:
+    {topic}. Summarize them in a few sentences.
+
+    {listed_articles}
+    """
+
+    num_articles: int
+    topic: str
+    articles: list[str]
+
+    @property
+    def listed_articles(self) -> str:
+        return "\n".join(
+            [f"{i+1}. {article}" for i, article in enumerate(self.articles)]
+        )
+```
+
+Another change from the soccer example is how we query our context, in this case `articles`. Instead of calculating the vector similarities locally, we call `Index.query()` on our Pinecone index to let them efficiently do the hard work for us. Once the articles are retrieved, we take their indices and search them up in our `pd.Dataframe` to match them to the articles, and let Mirascope take care of the rest.
+
+```python
+def summarize_news(query: str, num_articles: int) -> str:
+    """Summarizes 2004 news about retrieved context relevant to query."""
+    query_embedding = (
+        chat.client.embeddings.create(model=EMBEDDINGS_MODEL, input=[query])
+        .data[0]
+        .embedding
+    )
+    query_response = index.query(
+        namespace="articles", vector=query_embedding, top_k=num_articles
+    )
+    relevant_article_indices = [
+        int(article["id"]) for article in query_response["matches"]
+    ]
+    relevant_articles = df.loc[relevant_article_indices][TEXT_COLUMN].tolist()
+    prompt = NewsPrompt(
+        num_articles=num_articles, topic=query, articles=relevant_articles
+    )
+
+    return str(chat.create(prompt))
+
+
+print(summarize_news("teams and players going through trouble in soccer", 3))
```

Comment:
move NewsPrompt out of `summarize_news`
rename summarize_news to relevant_articles
print
```
NewsPrompt(num_articles=3, topic="teams and players going through trouble in soccer", articles=relevant_articles())
```